File: problem-sets/leetcode-weekly-contest/20240331/100240-最小化曼哈顿距离.py
from typing import List
from sortedcontainers import SortedList

class Solution:
    def minimumDistance(self, points: List[List[int]]) -> int:
        n = len(points)
        save = [[0, 0] for i in range(n)]
        for i, point in enumerate(points):
            save[i][0] = point[0] + point[1]
            save[i][1] = point[0] - point[1]

        sum_, diff_ = SortedList(), SortedList()
        for i in range(n):
            sum_.add(save[i][0])
            diff_.add(save[i][1])

        ret = 1e9
        for i in range(n):
            sum_.remove(save[i][0])
            diff_.remove(save[i][1])
            dist = max(sum_[-1] - sum_[0], diff_[-1] - diff_[0])
            ret = min(ret, dist)
            sum_.add(save[i][0])
            diff_.add(save[i][1])
            
        return ret

File: problem-sets/leetcode-weekly-contest/20240331/100240-最小化曼哈顿距离.cpp
#include <array>
#include <set>
#include <vector>

using std::array;
using std::multiset;
using std::vector;

class Solution
{
public:
  int minimumDistance(vector<vector<int>> &points)
  {
    int n = points.size();
    vector<array<int, 2>> a(n);
    for (int i = 0; i < n; ++i) {
      a[i][0] = points[i][0] + points[i][1];
      a[i][1] = points[i][0] - points[i][1];
    }

    multiset<int> sum, diff;
    for (int i = 0; i < n; ++i) {
      sum.insert(a[i][0]);
      diff.insert(a[i][1]);
    }

    int ret = 1e9;
    for (int k = 0; k < n; ++k) {
      sum.extract(a[k][0]);  // delete kth point
      diff.extract(a[k][1]);
      int cur = std::max(*sum.rbegin() - *sum.begin(), *diff.rbegin() - *diff.begin());
      ret     = std::min(ret, cur);
      sum.insert(a[k][0]);  // restore kth point
      diff.insert(a[k][1]);
    }
    return ret;
  }
};